### PR TITLE
nodes/log: fail early on invalid log path

### PR DESCRIPTION
**Pull request by enj**
`http.ServeFile` performs the `..` check already.  Validate the path explicitly instead of relying on that behavior.  Getting a particular log or all logs is the same authorization check, so this change simply makes extra sure that we never expose anything outside of the `/var/log` directory.

/kind cleanup
/milestone v1.28
/triage accepted

```release-note
NONE
```

### CODE REVIEW

**Comment by moficodes** [comment]
@enj @SergeyKanzhelev 

1.28 Bug Triage Shadow here.

Just checking-in, if this is still on track for the 1.28 release?

**Comment by furkatgofurov7** [comment]
@SergeyKanzhelev hey, Bug Triage Team Lead here!

Can you please help with reviewing this PR? 

Also, please be aware the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about less than 2 weeks from now).

**Comment by SergeyKanzhelev** [comment]
> `http.ServeFile` performs the `..` check already. Validate the path explicitly instead of relying on that behavior. Getting a particular log or all logs is the same authorization check, so this change simply makes extra sure that we never expose anything outside of the `/var/log` directory.

I am missing context here. Why do we want to fail earlier?

File: pkg/routes/logs.go [hunk 1]
```diff
@@ -68,3 +72,27 @@ func logFileNameIsTooLong(filePath string) bool {
...
+		return "", fmt.Errorf("invalid path: %q", logPath)
+	}
+	if logPath == "" || logPath == "/" {
+		return "", fmt.Errorf("empty path: %q", logPath)
+	}
+	// we ensure that the prefix ends in '/' in below, so skip any leading '/' in the log path now
+	startIndex := 0
+	if logPath[0] == '/' {
+		startIndex = 1
+	}
```
Comment (marquiz) [inline pkg/routes/logs.go:96]: Shouldn't we trim all leading slashes? Possibly run `path.Clean()`?

**Review by marquiz — COMMENTED** [review]
wdyt @SergeyKanzhelev ?

File: pkg/routes/logs.go [hunk 2]
```diff
@@ -68,3 +72,27 @@ func logFileNameIsTooLong(filePath string) bool {
...
+	if logPath == ".." ||
+		strings.HasPrefix(logPath, "../") ||
+		strings.HasSuffix(logPath, "/..") ||
+		strings.Contains(logPath, "/../") {
+		return "", fmt.Errorf("invalid path: %q", logPath)
+	}
+	if logPath == "." ||
+		strings.HasPrefix(logPath, "./") ||
+		strings.HasSuffix(logPath, "/.") ||
+		strings.Contains(logPath, "/./") {
```
Comment (SergeyKanzhelev) [inline pkg/routes/logs.go:86]: why is this bad? Just a no-op

File: pkg/routes/logs.go (hunk 1 above)
Comment (SergeyKanzhelev) [reply pkg/routes/logs.go:96]: +1 on `path.Clean()`. Join, than `path.Clean()`, than check that prefix is still the same

File: pkg/routes/logs.go [hunk 3]
```diff
@@ -68,3 +72,27 @@ func logFileNameIsTooLong(filePath string) bool {
...
+	}
+	if logPath == "" || logPath == "/" {
+		return "", fmt.Errorf("empty path: %q", logPath)
+	}
+	// we ensure that the prefix ends in '/' in below, so skip any leading '/' in the log path now
+	startIndex := 0
+	if logPath[0] == '/' {
+		startIndex = 1
+	}
+	return "/var/log/" + logPath[startIndex:], nil
```
Comment (SergeyKanzhelev) [inline pkg/routes/logs.go:97]: why not `path.Join`?

File: pkg/routes/logs.go [hunk 4]
```diff
@@ -41,8 +42,11 @@ func (l Logs) Install(c *restful.Container) {
 }
 
 func logFileHandler(req *restful.Request, resp *restful.Response) {
-	logdir := "/var/log"
-	actual := path.Join(logdir, req.PathParameter("logpath"))
```
Comment (SergeyKanzhelev) [inline pkg/routes/logs.go:45]: not related to this PR. But nevertheless, does it work for Windows?

**Comment by moficodes** [comment]
Hello @enj  ! 👋🏼
Bug triage for the 1.29 release cycle is here!
This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 2 weeks from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.

As the issue is tagged for 1.29, is it still planned for that release?

**Comment by moficodes** [comment]
Hello @enj ! 👋🏼
Bug triage for the 1.29 release cycle is here!
This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is this week. We want to make sure that every PR has a chance to be merged on time.

As the issue is tagged for 1.29, is it still planned for that release?

**Comment by hailkomputer** [comment]
Hello @enj and @SergeyKanzhelev, as we are past in the code freeze, I am clearing the milestone for the v1.29. If you think this will be worked on and needed in the next release, feel free to add the milestone for the next candidate.

/milestone clear

**Comment by bart0sh** [comment]
@enj please address remaining comments, thanks.

**Comment by enj** [comment]
@bart0sh I don't expect to get to this in the next 6 months.  Anyone is free to take it.

**Comment by bart0sh** [comment]
OK, thanks. Let's wait a couple of days. If nobody is willing to take it, let's close it.

**Pull request closed without merging**